client/src/connection/ssh/index.ts

```diff
@@ -0,0 +1,196 @@
+      }
+      stream = s;
+      if (!stream) {
+        reject?.(err);
+        return;
+      }
+
+      stream
+        .on("close", () => {
+          onLog = undefined;
+          stream = undefined;
+          resolve = undefined;
+          reject = undefined;
+          logs = [];
+          conn.end();
+        })
+        .on("data", (data: Buffer) => {
+          const output = data.toString().trimEnd();
+          if (onLog) {
+            setTimer();
+            output.split(/\n|\r\n/).forEach((line) => {
+              if (!line) {
+                return;
+              }
+              if (line.endsWith(endCode)) {
+                // run completed
+                clearTimer();
+                getResult();
+              }
+              if (!(line.endsWith("?") || line.endsWith(">"))) {
+                html5FileName =
+                  line.match(/NOTE: .+ HTML5.* Body .+: (.+)\.htm/)?.[1] ??
+                  html5FileName;
+                onLog?.([{ type: "normal", line }]);
+              }
+            });
+          } else {
+            logs.push(output);
+            if (output.endsWith("?")) {
+              clearTimer();
+              resolve?.();
+            }
+          }
+        });
+
+      const resolvedEnv: string[] = [
+        '_JAVA_OPTIONS="-Djava.awt.headless=true"',
+      ];
+      const execArgs: string = resolvedEnv.join(" ");
+
+      const resolvedSasOpts: string[] = [
+        "-nodms",
+        "-terminal",
+        "-nosyntaxcheck",
+      ];
+
+      if (config.sasOptions) {
+        resolvedSasOpts.push(...config.sasOptions);
+      }
+      const execSasOpts: string = resolvedSasOpts.join(" ");
+
+      stream.write(`${execArgs} ${config.saspath} ${execSasOpts} \n`);
+    });
+  })
+  .on("error", (err) => {
+    clearTimer();
+    reject?.(err);
+    return;
+  });
+
+function setTimer() {
+  clearTimer();
+  timer = setTimeout(() => {
+    reject?.(new Error("Time out"));
+    timer = undefined;
+  }, 10000);
+}
+
+function clearTimer() {
+  timer && clearTimeout(timer);
+  timer = undefined;
+}
+
+function getResult() {
+  const runResult: RunResult = {};
+  if (!html5FileName) {
+    resolve?.(runResult);
+    return;
+  }
+  let fileContents = "";
+  conn.exec(`cat ${html5FileName}.htm`, (err: Error, s: ClientChannel) => {
+    if (err) {
+      reject?.(err);
+      return;
+    }
+
+    s.on("data", (data) => {
+      fileContents += data.toString().trimEnd();
+    }).on("close", (code) => {
+      const rc = code as number;
+
+      if (rc === 0) {
+        //Make sure that the html has a valid body
+        //TODO: should this be refactored into a shared location?
+        if (fileContents.search('<*id="IDX*.+">') !== -1) {
+          runResult.html5 = fileContents;
+          runResult.title = "Result";
+        }
+      }
+      resolve?.(runResult);
+    });
+  });
+}
+
+function setup(): Promise<void> {
+  return new Promise((_resolve, _reject) => {
+    resolve = _resolve;
```

Comment:
Yea, we need to be able to reassign these at the module level because they could have different behaviors attached when setup is called versus when code executions happen, with needing access to the right one at the right time during the "on data" events for the stream. 

Definitely open to alternative ideas on how we might implement this.